temporalio/client.py

```diff
@@ -1984,114 +2095,1649 @@ async def __anext__(self) -> temporalio.api.history.v1.HistoryEvent:
+
+    jitter: Optional[timedelta] = None
+    """Jitter to apply each action.
+
+    An action's scheduled time will be incremented by a random value between 0
+    and this value if present (but not past the next schedule).
+    """
+
+    time_zone_name: Optional[str] = None
+    """IANA time zone name, for example ``US/Central``."""
+
+    @staticmethod
+    def _from_proto(spec: temporalio.api.schedule.v1.ScheduleSpec) -> ScheduleSpec:
+        return ScheduleSpec(
+            calendars=[
+                ScheduleCalendarSpec._from_proto(c) for c in spec.structured_calendar
+            ],
+            intervals=[ScheduleIntervalSpec._from_proto(i) for i in spec.interval],
+            cron_expressions=spec.cron_string,
+            skip=[
+                ScheduleCalendarSpec._from_proto(c)
+                for c in spec.exclude_structured_calendar
+            ],
+            start_at=spec.start_time.ToDatetime()
+            if spec.HasField("start_time")
+            else None,
+            end_at=spec.end_time.ToDatetime() if spec.HasField("end_time") else None,
+            jitter=spec.jitter.ToTimedelta() if spec.HasField("jitter") else None,
+            time_zone_name=spec.timezone_name or None,
+        )
+
+    def _to_proto(self) -> temporalio.api.schedule.v1.ScheduleSpec:
+        start_time: Optional[google.protobuf.timestamp_pb2.Timestamp] = None
+        if self.start_at:
+            start_time = google.protobuf.timestamp_pb2.Timestamp()
+            start_time.FromDatetime(self.start_at)
+        end_time: Optional[google.protobuf.timestamp_pb2.Timestamp] = None
+        if self.end_at:
+            end_time = google.protobuf.timestamp_pb2.Timestamp()
+            end_time.FromDatetime(self.end_at)
+        jitter: Optional[google.protobuf.duration_pb2.Duration] = None
+        if self.jitter:
+            jitter = google.protobuf.duration_pb2.Duration()
+            jitter.FromTimedelta(self.jitter)
+        return temporalio.api.schedule.v1.ScheduleSpec(
+            structured_calendar=[cal._to_proto() for cal in self.calendars],
+            cron_string=self.cron_expressions,
+            interval=[i._to_proto() for i in self.intervals],
+            exclude_structured_calendar=[cal._to_proto() for cal in self.skip],
+            start_time=start_time,
+            end_time=end_time,
+            jitter=jitter,
+            timezone_name=self.time_zone_name or "",
+        )
+
+
+@dataclass
+class ScheduleCalendarSpec:
+    """Specification relative to calendar time when to run an action.
+
+    A timestamp matches if at least one range of each field matches except for
+    year. If year is missing, that means all years match. For all fields besides
+    year, at least one range must be present to match anything.
+
+    .. warning::
+        Schedules are an experimental feature.
+    """
+
+    second: Sequence[ScheduleRange] = dataclasses.field(default_factory=list)
```

Comment:
These defaults are not OK. TIL that except for year, if these are empty, this will never match. I am still checking whether server errors. We have two options:

1. Do not default these fields requiring users to set them
2. Default them to `[ScheduleRange(0)]`

Reply:
Fixed by making the range immutable and setting default ranges.

Decided not to have an "any" and "all" static shortcut range sentinels because we want symmetry with input and output from describe. Also people that usually want these values would just leave the fields unset and thereby use their defaults. If someone wants to change from some fixed value to "any" or "all", hand-typing the range is not unreasonable.